### ISSUE TITLE
Add mean and var operation for Nested Tensors

### ISSUE THREAD

**Issue by mjdxyr** [pytorch/pytorch#138831]
### 🚀 The feature, motivation and pitch

For torch.Tesnor it is easy to compute mean and var, but I can not find ways to compute mean and var for Nested Tensors. Nested Tensors support layer_norm operation which include mean and var operation. Is there any ways to compute mean fo Nested Tensors?
```
x = torch.randn(1, 192)
y = torch.randn(10, 192)
nested = torch.nested.nested_tensor([x, y])
nested.mean(dim=-1)#Don't support
layer_norm(nested)#support
```

**Comment by jbschlosser** [comment]
Hey thanks for the request @mjdxyr! I'll suggest utilizing jagged layout nested tensors (AKA NJTs) as they are the future of nested tensors, with the older version of nested tensors being unsupported.

```python
import torch

x = torch.randn(1, 192)
y = torch.randn(10, 192)
nested = torch.nested.nested_tensor([x, y], layout=torch.jagged)
print(nested.shape)  # torch.Size([2, j1, 192])
output = nested.mean(dim=-1, keepdim=True)
print(output.shape)  # torch.Size([2, j1, 1])
```

Op support for NJTs should be much better, and I'm working hard to close any op coverage gaps. We support certain reductions over the jagged dimension, including sum and mean. Note in the example above that `keepdim=True` is required but I can relax this restriction - PR forthcoming shortly.

**Comment by mjdxyr** [comment]
> Hey thanks for the request @mjdxyr! I'll suggest utilizing jagged layout nested tensors (AKA NJTs) as they are the future of nested tensors, with the older version of nested tensors being unsupported.
> 
> ```python
> import torch
> 
> x = torch.randn(1, 192)
> y = torch.randn(10, 192)
> nested = torch.nested.nested_tensor([x, y], layout=torch.jagged)
> print(nested.shape)  # torch.Size([2, j1, 192])
> output = nested.mean(dim=-1, keepdim=True)
> print(output.shape)  # torch.Size([2, j1, 1])
> ```
> 
> Op support for NJTs should be much better, and I'm working hard to close any op coverage gaps. We support certain reductions over the jagged dimension, including sum and mean. Note in the example above that `keepdim=True` is required but I can relax this restriction - PR forthcoming shortly.

It works for model forward but raise RuntimError for backward.
```
/opt/conda/conda-bld/pytorch_1728945388038/work/build/aten/src/ATen/RegisterCompositeExplicitAutograd.cpp:2316: SymIntArrayRef expected to contain only concrete integers
  File "/root/Model_NestedTensor/run.py", line 89, in <module>
    loss.backward()
RuntimeError: /opt/conda/conda-bld/pytorch_1728945388038/work/build/aten/src/ATen/RegisterCompositeExplicitAutograd.cpp:2316: SymIntArrayRef expected to contain only concrete integers
```

**Comment by jbschlosser** [comment]
Fixes for mean here: #139317

Still need `var`, will add that in a follow-up.

**Comment by dquintanatorres** [comment]
Hi @jbschlosser, thanks for driving the work on reduction-op support for NJTs! I’m testing against PyTorch 2.9.0.dev20250804+cu129 and am hitting autograd errors when I reduce along the jagged dimension:

- With sum (`torch.sum(x, dim=1, keepdim=False)`):
```
File "...../.venv/lib/python3.12/site-packages/torch/nested/_internal/nested_tensor.py", line 348, in __torch_dispatch__
    raise NotImplementedError(func)
NotImplementedError: aten._nested_sum_backward.default
```
- With mean (`torch.mean(x, dim=1, keepdim=False)`):
```
File "...../.venv/lib/python3.12/site-packages/torch/autograd/graph.py", line 829, in _engine_run_backward
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: /pytorch/build/aten/src/ATen/RegisterCompositeExplicitAutograd_0.cpp:5016: SymIntArrayRef expected to contain only concrete integers
```

Do you know if backward support for these reductions is still in progress, or is there a workaround I could try in the meantime?

Thanks again!

**Comment by dquintanatorres** [comment]
That's unfortunate to hear. This was just a one-off requirement for my project, and I don't think I currently have the bandwidth or expertise to contribute effectively to this issue. In the meantime, I've resorted to converting NJTs to regular tensors with sensible padding as follows:

- Sum:
```
x = x.to_padded_tensor(padding=0.0)
x.sum(dim=1, keepdim=False)
```
- Mean:
```
x = x.to_padded_tensor(padding=float("nan")) 
x.nanmean(dim=1, keepdim=False)
```

If you have any suggestions for a better workaround, I'd be happy to hear your thoughts!

**Comment by jbschlosser** [comment]
@dquintanatorres yep I think converting to padded as you're doing is the right move, although of course you'll take on the conversion overhead and downsides of operating on padding in general.
